docs/src/arrays.md

```diff
@@ -0,0 +1,440 @@
+
+For both forward- and reverse-mode rules for matrices, the first step was to write down the differential identities, which followed directly from the scalar differential identities.
+This approach follows for arrays, but it's easier to work in component form.
+Consider the following function
+
+```julia
+C = sum(abs2, A::Array{<:RealOrComplex,3}; dims=2)::Array{<:Real,3}
+```
+
+which we write as
+
+$$C_{i1k} = \sum_{j} |A_{ijk}|^2 = \sum_{j} \operatorname{real}(\operatorname{conj}(A_{ijk}) A_{ijk})$$
+
+The differential identity is
+
+```math
+\begin{align*}
+    dC_{i1k} &= \sum_j \operatorname{real}( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
+                            \operatorname{conj}(A_{ijk}) ~dA_{ijk} ) \\
+             &= \sum_j \operatorname{real}(
+                    \operatorname{conj}\left(
+                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+                    \right) +
+                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+                )\\
+             &= \sum_j 2 \operatorname{real}( \operatorname{conj}(A_{ijk}) ~dA_{ijk} )
+\end{align*}
+```
+
+where in the last step we have used the fact that for all real $a$ and $b$,
+
+$$(a + i b) + \operatorname{conj}(a + i b) = (a + i b) + (a - i b) = 2 a = 2 \operatorname{real} (a + i b).$$
+
+The `frule` can be implemented generally as
+
+```julia
+function frule(
+    (_, _, ΔA),
+    ::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex};
+    dims = :,
+)
+    C = sum(abs2, A; dims = dims)
+    ∂C = sum(2 .* real.(conj.(A) .* ΔA); dims = dims)
+    return (C, ∂C)
+end
+```
+
+We can now derive the reverse-mode rule.
+The array form of the desired identity will be
+
+```math
+ds = \operatorname{real} \left( \sum_{ik}  \operatorname{conj}(\overline{C}_{i1k}) ~dC_{i1k} \right)
+   = \operatorname{real} \left( \sum_{ijk} \operatorname{conj}(\overline{A}_{ijk}) ~dA_{ijk} \right)
+```
+
+We plug the differential identity into the middle expression to get
+
+```math
+\begin{align*}
+    ds &= \operatorname{real} \left(\sum_{ijk}
+                  \operatorname{conj}(\overline{C}_{i1k})
+                  2 \operatorname{real}(\operatorname{conj}(A_{ijk}) ~dA_{ijk})
+              \right) \\
+       &= \operatorname{real} \left( \sum_{ijk}
+              2 \operatorname{real}(\overline{C}_{i1k})
+              \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+          \right).
+\end{align*}
+```
+
+We now solve for $\overline{A}$:
+
+```math
+\overline{A}_{ijk} = \operatorname{conj}(
+                         2 \operatorname{real}( \overline{C}_{i1k} )
+                         \operatorname{conj}(A_{ijk})
+                     )
+                   = 2\operatorname{real}( \overline{C}_{i1k} ) A_{ijk}
+```
+
+Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:
+
+```julia
+function rrule(::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex}; dims = :)
```

Comment:
same question re `dims`?

Reply:
In this case I wanted to derive a generic rule but writing that out in math notation is horrific, so I instead derived for an arbitrary choice of indices and then stated that it's generic. I added some more text (hopefully) making that clearer.